core/src/main/java/nl/nn/adapterframework/extensions/kafka/KafkaListener.java

```diff
@@ -0,0 +1,145 @@
+/*
+   Copyright 2023 WeAreFrank!
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+package nl.nn.adapterframework.extensions.kafka;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Pattern;
+
+import javax.annotation.Nonnull;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.common.serialization.ByteArrayDeserializer;
+import org.apache.kafka.common.serialization.StringDeserializer;
+
+import lombok.AccessLevel;
+import lombok.Setter;
+import nl.nn.adapterframework.configuration.ConfigurationException;
+import nl.nn.adapterframework.core.IMessageHandler;
+import nl.nn.adapterframework.core.IPushingListener;
+import nl.nn.adapterframework.core.IbisExceptionListener;
+import nl.nn.adapterframework.core.ListenerException;
+import nl.nn.adapterframework.core.PipeLineResult;
+import nl.nn.adapterframework.core.PipeLineSession;
+import nl.nn.adapterframework.receivers.RawMessageWrapper;
+import nl.nn.adapterframework.stream.Message;
+
+public class KafkaListener extends KafkaFacade implements IPushingListener<ConsumerRecord<String, byte[]>> {
+	private IMessageHandler<ConsumerRecord<String, byte[]>> messageHandler;
+	private IbisExceptionListener ibisExceptionListener;
+	//setter is for testing purposes only.
+	private @Setter(AccessLevel.PACKAGE) Consumer<String, byte[]> consumer;
+	/** The group id of the consumer */
+	private @Setter String groupId;
+	/** Whether to start reading from the beginning of the topic. */
+	private @Setter boolean fromBeginning = false;
+	/** How often to check for new topics when using Patterns. (in MS) */
+	private @Setter int patternRecheckInterval = 5000;
+	/** The topics to listen to, separated by `,`. Wildcards are supported with `example.*`. */
+	private @Setter String topics;
+	/** Kafka internal poll timeout (in MS) */
+	private @Setter int pollTimeout = 100;
+	private KafkaInternalConsumer poller;
```

Comment:
Je bent nu een polling implementatie aan het maken in een `IPushingListener`, heb je ook overwogen hier een `IPullingListener` voor te gebruiken?

Reply:
Ik heb 'm omgebouwd, maar ik weet niet of hij goed met de threads om kan gaan. Is daar een simpele manier voor om dat te testen?

Reply:
Lastig. Je zou een integration-test moeten maken met een Kafka docker image en een Frank! configuratie die dit test. Dat zou sowieso een heel goed idee zijn.

Als er meerdere poll-threads zijn, dan wordt dezelfde listener-instance gedeeld over alle poll-threads. Dus of het goed gaat of niet, hangt af van in hoeverre de KafkaConsumer thread-safe is. Dat kan ik zo snel niet zien in de API docs, maar ik verwacht dat dat wel ergens gedocumenteerd zal zijn.

Reply:
Met een beetje zoeken in de Kafka library source heb ik gevonden dat het niet thread safe is. Ik ga later verder uitzoeken wat ik hiermee moet doen.

Reply:
Ik heb de listener nu geforceert op 1 thread, maar dat is niet heel netjes. Ik denk niet dat hier een andere optie voor is...

Reply:
Lock objects gebruiken?